source/common/upstream/upstream_impl.cc

```diff
@@ -87,6 +87,65 @@ HostsPerLocalityImpl::filter(std::function<bool(const Host&)> predicate) const {
   return shared_filtered_clone;
 }
 
+void HostSetImpl::updateHosts(HostVectorConstSharedPtr hosts,
+                              HostVectorConstSharedPtr healthy_hosts,
+                              HostsPerLocalityConstSharedPtr hosts_per_locality,
+                              HostsPerLocalityConstSharedPtr healthy_hosts_per_locality,
+                              LocalityWeightsConstSharedPtr locality_weights,
+                              const HostVector& hosts_added, const HostVector& hosts_removed) {
+  hosts_ = std::move(hosts);
+  healthy_hosts_ = std::move(healthy_hosts);
+  hosts_per_locality_ = std::move(hosts_per_locality);
+  healthy_hosts_per_locality_ = std::move(healthy_hosts_per_locality);
+  locality_weights_ = std::move(locality_weights);
+  // Rebuild the locality scheduler.
+  if (hosts_per_locality_ != nullptr && locality_weights_ != nullptr &&
+      !locality_weights_->empty()) {
+    locality_scheduler_ = std::make_unique<EdfScheduler<LocalityEntry>>();
+    locality_entries_.clear();
+    for (uint32_t i = 0; i < hosts_per_locality_->get().size(); ++i) {
+      const double effective_weight = effectiveLocalityWeight(i);
+      if (effective_weight > 0) {
+        locality_entries_.emplace_back(std::make_shared<LocalityEntry>(i, effective_weight));
+        locality_scheduler_->add(effective_weight, locality_entries_.back());
+      }
+    }
+  } else {
+    locality_scheduler_ = nullptr;
+  }
+  runUpdateCallbacks(hosts_added, hosts_removed);
+}
+
+absl::optional<uint32_t> HostSetImpl::chooseLocality() {
+  if (locality_scheduler_ == nullptr) {
+    return {};
+  }
+  const std::shared_ptr<LocalityEntry> locality = locality_scheduler_->pick();
+  // We don't build a schedule if there are no weighted localities, so we should always succeed.
+  ASSERT(locality != nullptr);
+  // If we picked it before, its weight must have been positive.
+  ASSERT(locality->effective_weight_ > 0);
+  locality_scheduler_->add(locality->effective_weight_, locality);
```

Comment:
Yeah, this is actually deliberate, as we use it in host-level picking for the lazy weight updates. I'll leave it as is, since as discussed above, we'd like to do this at the locality level as well.